transport/transport.go

```diff
@@ -506,3 +507,50 @@ func wait(ctx context.Context, closing <-chan struct{}, proceed <-chan int) (int
 		return i, nil
 	}
 }
+
+const (
+	spaceByte   = byte(int(' '))
+	tildaByte   = byte(int('~'))
+	percentByte = byte(int('%'))
+)
+
+// matching https://github.com/grpc/grpc-java/pull/1517/files
+func grpcMessageEncode(grpcMessage string) string {
+	if grpcMessage == "" {
+		return ""
+	}
+	var buffer bytes.Buffer
+	for _, c := range []byte(grpcMessage) {
+		if c >= spaceByte && c < tildaByte && c != percentByte {
+			_ = buffer.WriteByte(c)
+		} else {
+			_, _ = buffer.WriteString(fmt.Sprintf("%%%02X", c))
+		}
+	}
+	return buffer.String()
+}
+
+// matching https://github.com/grpc/grpc-java/pull/1517/files
+func grpcMessageDecode(encodedGrpcMessage string) string {
+	if encodedGrpcMessage == "" {
+		return ""
+	}
+	var buffer bytes.Buffer
+	data := []byte(encodedGrpcMessage)
```

Comment:
don't do this conversion. unnecessary allocation. iterate over the bytes instead.

but do a pass over it first to see if it's valid before you do any allocations.